File: bindings/haskell/src/ProvenServers/Loadbalancer.hs
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
--
-- | Load Balancer types for the proven-servers ABI.
--
-- All tag values match the Idris2 ABI discriminants exactly.

module ProvenServers.Loadbalancer
  (
    Algorithm(..)
  , algorithmToTag
  , algorithmFromTag
  , HealthCheckType(..)
  , healthCheckTypeToTag
  , healthCheckTypeFromTag
  , BackendState(..)
  , backendStateToTag
  , backendStateFromTag
  , canReceiveTraffic
  , SessionPersistence(..)
  , sessionPersistenceToTag
  , sessionPersistenceFromTag
  , LbProtocol(..)
  , lbProtocolToTag
  , lbProtocolFromTag
  ) where

import Data.Word (Word8)

-- ---------------------------------------------------------------------------
-- Algorithm
-- ---------------------------------------------------------------------------

-- | Load balancing algorithms.
--
-- Tags 0-5 (6 constructors).
data Algorithm
  = RoundRobin  -- ^ RoundRobin (tag 0).
  | LeastConnections  -- ^ LeastConnections (tag 1).
  | IpHash  -- ^ IpHash (tag 2).
  | Random  -- ^ Random (tag 3).
  | WeightedRoundRobin  -- ^ WeightedRoundRobin (tag 4).
  | LeastResponseTime  -- ^ LeastResponseTime (tag 5).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'Algorithm' to its ABI tag value.
algorithmToTag :: Algorithm -> Word8
algorithmToTag = fromIntegral . fromEnum

-- | Decode a 'Algorithm' from its ABI tag value.
algorithmFromTag :: Word8 -> Maybe Algorithm
algorithmFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: Algorithm)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- HealthCheckType
-- ---------------------------------------------------------------------------

-- | Backend health check types.
--
-- Tags 0-3 (4 constructors).
data HealthCheckType
  = Http  -- ^ HTTP health check (tag 0).
  | Tcp  -- ^ TCP health check (tag 1).
  | Grpc  -- ^ gRPC health check (tag 2).
  | Script  -- ^ Script (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'HealthCheckType' to its ABI tag value.
healthCheckTypeToTag :: HealthCheckType -> Word8
healthCheckTypeToTag = fromIntegral . fromEnum

-- | Decode a 'HealthCheckType' from its ABI tag value.
healthCheckTypeFromTag :: Word8 -> Maybe HealthCheckType
healthCheckTypeFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: HealthCheckType)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- BackendState
-- ---------------------------------------------------------------------------

-- | Backend server states.
--
-- Tags 0-3 (4 constructors).
data BackendState
  = Healthy  -- ^ Healthy (tag 0).
  | Unhealthy  -- ^ Unhealthy (tag 1).
  | Draining  -- ^ Draining (tag 2).
  | Disabled  -- ^ Disabled (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'BackendState' to its ABI tag value.
backendStateToTag :: BackendState -> Word8
backendStateToTag = fromIntegral . fromEnum

-- | Decode a 'BackendState' from its ABI tag value.
backendStateFromTag :: Word8 -> Maybe BackendState
backendStateFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: BackendState)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- | Whether this backend can receive new connections.
canReceiveTraffic :: BackendState -> Bool
canReceiveTraffic Healthy = True
canReceiveTraffic _ = False

-- ---------------------------------------------------------------------------
-- SessionPersistence
-- ---------------------------------------------------------------------------

-- | Session persistence strategies.
--
-- Tags 0-3 (4 constructors).
data SessionPersistence
  = None  -- ^ None (tag 0).
  | Cookie  -- ^ Cookie (tag 1).
  | SourceIp  -- ^ Source IP affinity (tag 2).
  | Header  -- ^ Header (tag 3).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'SessionPersistence' to its ABI tag value.
sessionPersistenceToTag :: SessionPersistence -> Word8
sessionPersistenceToTag = fromIntegral . fromEnum

-- | Decode a 'SessionPersistence' from its ABI tag value.
sessionPersistenceFromTag :: Word8 -> Maybe SessionPersistence
sessionPersistenceFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: SessionPersistence)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

-- ---------------------------------------------------------------------------
-- LbProtocol
-- ---------------------------------------------------------------------------

-- | Load balancer protocols.
--
-- Tags 0-4 (5 constructors).
data LbProtocol
  = Http  -- ^ HTTP (tag 0).
  | Https  -- ^ HTTPS (tag 1).
  | Tcp  -- ^ TCP (tag 2).
  | Udp  -- ^ UDP (tag 3).
  | Grpc  -- ^ gRPC (tag 4).
  deriving (Show, Eq, Ord, Enum, Bounded)

-- | Convert a 'LbProtocol' to its ABI tag value.
lbProtocolToTag :: LbProtocol -> Word8
lbProtocolToTag = fromIntegral . fromEnum

-- | Decode a 'LbProtocol' from its ABI tag value.
lbProtocolFromTag :: Word8 -> Maybe LbProtocol
lbProtocolFromTag n
  | n <= fromIntegral (fromEnum (maxBound :: LbProtocol)) = Just (toEnum (fromIntegral n))
  | otherwise = Nothing

File: bindings/elixir/lib/proven_servers/loadbalancer.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Loadbalancer do
  @moduledoc """
  Load Balancer types for the proven-servers ABI.
  
  Formally verified load balancer types.
  Mirrors the Idris2 module `LoadbalancerABI.Types`.
  
  - `Algorithm` -- Load balancing algorithms.
  - `HealthCheckType` -- Backend health check types.
  - `BackendState` -- Backend server states.
  - `SessionPersistence` -- Session persistence strategies.
  - `LbProtocol` -- Load balancer protocols.
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  # ===========================================================================
  # Algorithm (tags 0-5)
  # ===========================================================================

  @typedoc """
  Algorithm types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type algorithm ::
          :round_robin
          | :least_connections
          | :ip_hash
          | :random
          | :weighted_round_robin
          | :least_response_time

  @algorithm_tags %{
    round_robin: 0,
    least_connections: 1,
    ip_hash: 2,
    random: 3,
    weighted_round_robin: 4,
    least_response_time: 5,
  }

  @tag_to_algorithm Map.new(@algorithm_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Algorithm` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..5, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Loadbalancer.algorithm_from_tag(0)
      {:ok, :round_robin}
  """
  @spec algorithm_from_tag(non_neg_integer()) :: {:ok, algorithm()} | :error
  def algorithm_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 5 do
    {:ok, Map.fetch!(@tag_to_algorithm, tag)}
  end

  def algorithm_from_tag(_tag), do: :error

  @doc """
  Encode a `Algorithm` to the C-ABI tag value.
  """
  @spec algorithm_to_tag(algorithm()) :: non_neg_integer()
  def algorithm_to_tag(val) when is_map_key(@algorithm_tags, val) do
    Map.fetch!(@algorithm_tags, val)
  end

  @doc """
  All `Algorithm` variants in tag order.
  """
  @spec all_algorithms() :: [algorithm()]
  def all_algorithms do
    [
      :round_robin, :least_connections, :ip_hash, :random, :weighted_round_robin,
      :least_response_time
    ]
  end

  # ===========================================================================
  # HealthCheckType (tags 0-3)
  # ===========================================================================

  @typedoc """
  HealthCheckType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type health_check_type :: :http | :tcp | :grpc | :script

  @health_check_type_tags %{
    http: 0,
    tcp: 1,
    grpc: 2,
    script: 3,
  }

  @tag_to_health_check_type Map.new(@health_check_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `HealthCheckType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Loadbalancer.health_check_type_from_tag(0)
      {:ok, :http}
  """
  @spec health_check_type_from_tag(non_neg_integer()) :: {:ok, health_check_type()} | :error
  def health_check_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_health_check_type, tag)}
  end

  def health_check_type_from_tag(_tag), do: :error

  @doc """
  Encode a `HealthCheckType` to the C-ABI tag value.
  """
  @spec health_check_type_to_tag(health_check_type()) :: non_neg_integer()
  def health_check_type_to_tag(val) when is_map_key(@health_check_type_tags, val) do
    Map.fetch!(@health_check_type_tags, val)
  end

  @doc """
  All `HealthCheckType` variants in tag order.
  """
  @spec all_health_check_types() :: [health_check_type()]
  def all_health_check_types, do: [:http, :tcp, :grpc, :script]

  # ===========================================================================
  # BackendState (tags 0-3)
  # ===========================================================================

  @typedoc """
  BackendState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type backend_state :: :healthy | :unhealthy | :draining | :disabled

  @backend_state_tags %{
    healthy: 0,
    unhealthy: 1,
    draining: 2,
    disabled: 3,
  }

  @tag_to_backend_state Map.new(@backend_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `BackendState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Loadbalancer.backend_state_from_tag(0)
      {:ok, :healthy}
  """
  @spec backend_state_from_tag(non_neg_integer()) :: {:ok, backend_state()} | :error
  def backend_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_backend_state, tag)}
  end

  def backend_state_from_tag(_tag), do: :error

  @doc """
  Encode a `BackendState` to the C-ABI tag value.
  """
  @spec backend_state_to_tag(backend_state()) :: non_neg_integer()
  def backend_state_to_tag(val) when is_map_key(@backend_state_tags, val) do
    Map.fetch!(@backend_state_tags, val)
  end

  @doc """
  All `BackendState` variants in tag order.
  """
  @spec all_backend_states() :: [backend_state()]
  def all_backend_states, do: [:healthy, :unhealthy, :draining, :disabled]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this backend can receive new connections.
  """
  @spec can_receive_traffic?(backend_state()) :: boolean()
  def can_receive_traffic?(val) when val in [:healthy], do: true
  def can_receive_traffic?(_val), do: false

  # ===========================================================================
  # SessionPersistence (tags 0-3)
  # ===========================================================================

  @typedoc """
  SessionPersistence types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type session_persistence :: :none | :cookie | :source_ip | :header

  @session_persistence_tags %{
    none: 0,
    cookie: 1,
    source_ip: 2,
    header: 3,
  }

  @tag_to_session_persistence Map.new(@session_persistence_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `SessionPersistence` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Loadbalancer.session_persistence_from_tag(0)
      {:ok, :none}
  """
  @spec session_persistence_from_tag(non_neg_integer()) :: {:ok, session_persistence()} | :error
  def session_persistence_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_session_persistence, tag)}
  end

  def session_persistence_from_tag(_tag), do: :error

  @doc """
  Encode a `SessionPersistence` to the C-ABI tag value.
  """
  @spec session_persistence_to_tag(session_persistence()) :: non_neg_integer()
  def session_persistence_to_tag(val) when is_map_key(@session_persistence_tags, val) do
    Map.fetch!(@session_persistence_tags, val)
  end

  @doc """
  All `SessionPersistence` variants in tag order.
  """
  @spec all_session_persistences() :: [session_persistence()]
  def all_session_persistences, do: [:none, :cookie, :source_ip, :header]

  # ===========================================================================
  # LbProtocol (tags 0-4)
  # ===========================================================================

  @typedoc """
  LbProtocol types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type lb_protocol :: :http | :https | :tcp | :udp | :grpc

  @lb_protocol_tags %{
    http: 0,
    https: 1,
    tcp: 2,
    udp: 3,
    grpc: 4,
  }

  @tag_to_lb_protocol Map.new(@lb_protocol_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `LbProtocol` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Loadbalancer.lb_protocol_from_tag(0)
      {:ok, :http}
  """
  @spec lb_protocol_from_tag(non_neg_integer()) :: {:ok, lb_protocol()} | :error
  def lb_protocol_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_lb_protocol, tag)}
  end

  def lb_protocol_from_tag(_tag), do: :error

  @doc """
  Encode a `LbProtocol` to the C-ABI tag value.
  """
  @spec lb_protocol_to_tag(lb_protocol()) :: non_neg_integer()
  def lb_protocol_to_tag(val) when is_map_key(@lb_protocol_tags, val) do
    Map.fetch!(@lb_protocol_tags, val)
  end

  @doc """
  All `LbProtocol` variants in tag order.
  """
  @spec all_lb_protocols() :: [lb_protocol()]
  def all_lb_protocols, do: [:http, :https, :tcp, :udp, :grpc]

end
